### DISCUSSION TITLE
Enqueuing a new job returns a handle that can be awaited

### DISCUSSION THREAD

**Discussion by maxcountryman** [maxcountryman/underway#43] [Ideas]
I think it's probably pretty useful to be able to know when a job is finished. Currently when you enqueue a new job it more or less disappears into the ether (you do get the first step's task ID back, so it's not entirely lost).

Alternatively if we were to return a future that represented the execution of the complete job that would give callers the ability to coordinate jobs. (Perhaps eventually Underway could even build more complex workflow orchestration around this.)

This could be in the form of a handle that also exposed other methods, both for state introspection (completed_steps, failures, etc), but also for operational things (cancel, delete, reset, etc).

@kirillsalykin thoughts on this?

**Comment by kirillsalykin** [comment]
hi @maxcountryman ,

sorry for being quite - got some issues with my computer

I dont see any scenarios when this can be useful (doesnt mean it is not useful tho),
if you need to wait for a completion of it - maybe it shouldnt be a background job?

**Reply by maxcountryman** [reply]
All good. :)

Coordination is where I think it's probably most useful. But I agree, I don't think that's currently the general idea here (perhaps later on we could have an idea of workflows which are composed of jobs and there it would be necessary). 

That said, I do think we need to change what `Job::enqueue` returns. It doesn't quite make sense to return the task ID. Or put another way: we have a job ID which we could return instead and that would allow us to find all tasks belonging to that job enqueue.

If we're going to do that we could wrap that in another type to make it easier to e.g. cancel a job:

```rust
let enqueued_job = my_job.enqueue(&()).await?;
enqueued_job.cancel().await?;
```

Alternatively we could have a `Job::cancel(job_id: JobId)`.

There's a few methods like this that we probably want, including e.g. `delete` and `resume`.

Curious if any of this resonates with the use cases you have in mind.

**Reply by kirillsalykin** [reply]
I am still a bit sceptical about returning a future, workflows sounds cool, but it assumes that it will run in one process, which is usually not the case.
I am totally in for `JobId`, or maybe `Job` (so we can have extended view), this can be indeed useful for `delete`/`resume`.

Another scenario I see (based on my exp) - is when you have dependent jobs (not steps, dependent jobs (from different queues) with explicit deps, I think i shared the way I've implemented it).
(of course it is not given that it will be implemented)

**Comment by maxcountryman** [comment]
>  I am still a bit sceptical about returning a future

I think we might not be understanding each other: I'm saying that it would implement `Future`. That's different than it being a future in the same way that say a function is. For example, consider the handle we return from `start`: that's also a future in that it implements `Future` and a similar approach seems reasonable to me here. After all, an enqueued job is quite literally a `Future` in the purest sense.

**Reply by kirillsalykin** [reply]
ah, so you can pull it to schedule task, got it

**Reply by kirillsalykin** [reply]
then yeah, totally makes sense
👍 
sorry for misunderstanding

**Reply by maxcountryman** [reply]
Ah still not quite, no. :) It's running already. The idea is that if you want to await it you can and if you do you'll know it was completed (or errored) when the future is no longer pending.

**Reply by kirillsalykin** [reply]
ah, then i am not sure why this would be needed.

1. `enqueue_job` -> Future<JobResult>
2. `await`ing  ` Future<JobResult>` will wait for Future (read Job) to complete

**Reply by kirillsalykin** [reply]
right?

**Reply by maxcountryman** [reply]
Maybe the best way to approach this is to start by returning a type that encapsulates the job ID: this is where we can provide methods like `cancel`, etc. (However, if we think those methods actually belong on `Job` itself, this design doesn't make sense.) Then later we could implement `Future` for this if that's useful. (I think it probably is useful: it's at least a semantic that fits `Future` as far as I understand.)

**Reply by maxcountryman** [reply]
`Job::enqueue` would return `EnqueueHandle` or whatever we want to call it which wraps the job ID belonging to that specific enqueue of the job.

It's possible, but not required, to implement `Future` for `EnqueueHandle`.


`EnqueueHandle` also provides methods like `cancel`, `delete`, `resume`, etc.

**Reply by kirillsalykin** [reply]
got it, looks clear to me
i think it totally makes sense

**Reply by maxcountryman** [reply]
Thanks for taking a look! That's the next thing I'd like to do before cutting a new release. I think with that, we can start to plan for `v0.1.0` which should also mark semver stability.
